### ISSUE TITLE
Feature request: `blockchain.transaction.subscribe` and `blockchain.transaction.unsubscribe`

### ISSUE THREAD

**Issue by monsterbitar** [cculianu/Fulcrum#32]
Currently, in order to get notified when a transaction is confirmed, you have to subscribe to all changes to the recipient address, then look for changes to the specific transaction itself. This is problematic, because when done, if you unsubscribe from the address you might unintentionally unsubscribe from address changes you previously subscribed to elsewhere.

A proper solution is to add transaction subscribe/unsubscribe methods similar to the address subscriptions, that return the status of the transaction (some hash of the transaction metadata). This would be safe(r) to subscribe and unsubscribe to, and also easier to understand and would keep code cleaner in both design and legibility.

**Comment by monsterbitar** [comment]
Correction, the `blockchain.headers.subscribe` is a better call to get a notification on change, and makes more sense than the address.

My bad, this is signficantly less important than I initially assumed.

**Comment by cculianu** [comment]
Hmm yeah. I see what you mean.  Yeah I guess getting notified of a header (e.g. a block) is probably the fastest thing right now with the current API -- then you can query the address's balance, history, list_unspent, etc.

As for tx sub/unsub.  Hmm.  I'll think about it.  That's not a bad idea -- I just think it would be some work to implement and test, so there's a resource challenge there.  I do this as a volunteer for free and all, so it's a passion project. 

I'll leave this issue open and maybe if others chime in that they need this feature for their application -- it's an idea on the table and can maybe happen some day.

I would hate to spend a week implementing it then nobody uses it or very few people use it... :/

**Comment by monsterbitar** [comment]
One benefit of a tx sub/unsub over block headers, is that if the tx gets dropped from the mempool, or if it gets double-spent, or if there's some CPFP metadata available, those could get exposed while waiting for a block header would not provide that data.

You're right in that this is neither urgent or very important, but over time as small issues keep stacking up I feel it might end up holding back new developers from joining the space - it's not very inspiring to see some of these "unclean" solutions.

For now, I'll work around it, and we'll see if others chime in later.

**Comment by cculianu** [comment]
> CPFP metadata available, 

You'd like to get notified on descendant tx's too? Or?

**Comment by monsterbitar** [comment]
The core requirement would be to know if the transaction gets confirmed, and if it is dropped from the mempool.

Being able to get notification when anything that may impact the transactions likelihood of being confirmed or dropped is beneficial (allows building better payment processing tools / risk estimation tools), but once a transaction is confirmed I think it makes more sense to track the individual UTXO's, rather than continuing to track the transaction.

With regards to descendants / CPFP - since the total accumulated fee for the transaction chain might impact a miners behaviour, it would be nice if that indeed triggered a status updated (and there was some way to get that information).

**Comment by cculianu** [comment]
interesting.  It might also be possible calculate an avg fee total for all tx's in the "package".  I could see the API offering a status update when descendent count changes and the status info would include some information like that so client code can "tell" if a TX looks like it is very generously rewarding miners versus not, etc. 

Of course right now everything pretty much confirms if it made it to the miner within the last 10 seconds or so.. so long as it can get relayed.

But that may always not be the case.

Interesting feature.  Will definitely keep it on the table as something that would be nice to have.

**Comment by monsterbitar** [comment]
Dagur made a nice list in the corresponding issue in electrscash, that might be good to be aware of: https://github.com/BitcoinUnlimited/ElectrsCash/issues/86

**Comment by monsterbitar** [comment]
I'd like to point out that readcash is currently asking BCHD to implement a similar feature to learn when/if transactions are evicted frm the mempool and why (as part of a much larger issue):

https://github.com/gcash/bchd/issues/404

**Comment by cculianu** [comment]
Ah thanks for that link. So.. the "competition" is already threatening to beat us!  But seriously thanks for the link.

**Comment by cculianu** [comment]
Hey Jonathan this might be related:

https://github.com/spesmilo/electrumx/pull/80

ElectrumX guys want to add `blockchain.outpoint.subscribe` (see the PR for the actual specification). It's not exactly `blockchain.transaction.subscribe` -- but it's.. sort of in the same ballpark, but not exactly what you asked for here.  Let me know your thoughts on that.  They need it for lightning.

**Comment by monsterbitar** [comment]
I wouldn't mind being able to subscribe to TXOs. If I can create a TX locally, subscribe to the TXO and then broadcast it, would I get similar behaviour to an address status? (that is, would I get notifcation for block inclusion as well as for when it gets spent? Could this be a reasonable entry-point to get notified of DS-proofs?)

**Comment by cculianu** [comment]
So it would be useful to you?  

Ok, just to answer your question(s) (I hope) -- so the full "life cycle"  for an `outpoint.subscribe` would be:

1. If it doesn't exist, you get notified with: `{}`
2. If it then later exists as an outpoint but is **unspent**, you get notified with: `{ "height" : xxxx }` (where height may be 0 or a confirmed height for the txo that now exists in a tx).
3. If it is later spent, you get notified: `{"height": xxxx, "spender_hash" : "xyz", "spender_height":  confirmed_height_0_for_mempool }`
4. (Unlikely) If its spending tx is reorged later out of existence, you get back the same notification as in (1) above.
5. (Unlikely) If its spending tx ends up as a different tx due to a double-spend, you get an updated (3) again with the correct info.

Note that we can "extend" this facility, potentially to also include dsproof information if available from the node... (for mempool-only tx's obviously).  Perhaps offer an additional key as an extension, "double-spends", or something.

---

Unrelated: I do want to add the original method you asked for, `transaction.subscribe` !  It does require a new index but is something we may want anyway regardless of this `outpoint.subscribe` business...

**Comment by monsterbitar** [comment]
Yes, I would very much use TXO subscription. The case I have is that I have a funding transaction for a smart contract, where I have some inputs (don't care about those), and multiple outputs, where I only care about tracking a specific one.

For my cashual wallet (that is still on ice), I also think it makes sense to track the UTXOs rather than addresses, but will need more time before I can get back to that project.

**Comment by cculianu** [comment]
Great to hear that!

Ok so I will do this method `blockchain.transaction.subscribe` as well as `blockchain.outpoint.subscribe`.  We will be implementing a superset of the BTC Electrum protocol 1.5, and `transaction.subscribe` will be a BCH extension.  I want to keep compatibility with Electrum fairly close since I feel that's a safe road to walk upon and ensures a solid future. We can always add extensions (such as this!).

But yes good to hear `outpoint.subscribe` may be useful for us as well.  This cinches it for me -- definitely will implement it now.

**Comment by cculianu** [comment]
(I'll let both you and @dagurval know when I actually get around to implementing these so I can get more feedback on the exact API)

**Comment by cculianu** [comment]
Hmm, this issue has seen no action. I can implement something simple easily now -- but maybe it's too simple -- however I feel it would actually cover most usecases:

`blockchain.transaction.subscribe` - takes a `txid` argument.  Always succeeds, immediately returns a single value which is one of the following:

- `null` - indicates the transaction is not in either the mempool or the blockchain. However the subscription will remain alive for at least 2? 10? 20? minutes (or 30? the exact amount of time is negotiable here -- can also be indefinite).  
- `"confirmed"` - indicates the transaction is known and already confirmed in the blockchain. The subscription will remain alive for at least 2? 10? 20? minutes (or again, up to debate -- can be indefinite).
- `"mempool"` - indicates the transaction is known and is in the mempool. The subscription will remain alive indefinitely until either the tx is lost (dropped from mempool) or is confirmed -- after which time perhaps an expiration timer will start and the subscription will be reaped if nothing changes on it during that time.

While the subscription is alive, if the tx has changed status (either has become confirmed, has become lost, or having been confirmed before, ended up back in the mempool again, or having been confirmed due to massive reorg, having become lost), then the client is notified of the change by receiving a notification with one of the above values:

Example notification:

`{"method":"blockchain.transaction.subscribe","params":["TXID_HEX_HERE","confirmed"]}`

If later on there is a reorg and it's back in mempool:

`{"method":"blockchain.transaction.subscribe","params":["TXID_HEX_HERE","mempool"]}`

Or, of later on it gets dropped:

`{"method":"blockchain.transaction.subscribe","params":["TXID_HEX_HERE",null]}`

Note that basically a subscription is up for cleanup if it has most recently given a `null` or `"confirmed"` reply. The motiviation behind the auto-cleanup is I am wary of clients that never unsubscribe from anything.... (Again, this is up for debate and can be indefinite always).

---

What do you think?

**Comment by monsterbitar** [comment]
Seems reasonable to me. I would like to also have the double-spend as a possible event in there, and I would prefer if the immidiate result of the call had the exact same structure as the subscription results (ie, include the TXID back in the result).

**Comment by cculianu** [comment]
Ih, I already did the double-spend as a separate subscription -- here is a brief synopsis:

`blockchain.transaction.dsproof.subscribe` - returns either `null` or a JSON object containing the double-spend proof + metadata.  You can subscribe to any tx and if it or any of its mempool parents has or gets a double-spend, you will get notified.  You also get notified if the double-spend state changes (it can chance say if your parent had a double-spend and so you also use the same proof, but then YOU also got a double spend, in which case your "best proof" becomes the proof for your tx).

I decided to do it as a separate event to keep things simple.  It made the implementation much simpler. And also server admins can elect to implement what they want or not want.  That and dsproof right now is beta, and only implemented via RPC in BCHN.

**Comment by monsterbitar** [comment]
Alright. I'm ok with subscribing to both dsproofs and transactions as separate if they's how you prefer it architected on your side.

**Comment by cculianu** [comment]
Yeah, awesome. Thanks: 👍

**Comment by cculianu** [comment]
Hey man so I ended up implementing this. It did require a new index but that index was needed anyway. See PR #75 

So the RPC API is slightly different than I first described it -- I realized it's more useful to return an actual block height rather than just `"confirmed"`.  Here is an *updated* synopsis.  I will of course release real documentation for this on this site: https://electrum-cash-protocol.readthedocs.io/en/latest/

I edited the description of this RPC in [my comment above](https://github.com/cculianu/Fulcrum/issues/32#issuecomment-784583714).
